File: android-app/lib/core/models/transaction_card_display.dart
import 'transaction_kind.dart';
import '../transaction_time.dart';

enum TransactionCardDisplayMode {
  contentFirst,
  categoryFirst,
}

extension TransactionCardDisplayModeX on TransactionCardDisplayMode {
  String get storageKey => switch (this) {
        TransactionCardDisplayMode.contentFirst => 'content_first',
        TransactionCardDisplayMode.categoryFirst => 'category_first',
      };

  String get label => switch (this) {
        TransactionCardDisplayMode.contentFirst => '内容优先',
        TransactionCardDisplayMode.categoryFirst => '分类优先',
      };

  static TransactionCardDisplayMode fromStorage(String? value) {
    for (final mode in TransactionCardDisplayMode.values) {
      if (mode.storageKey == value) return mode;
    }
    return TransactionCardDisplayMode.contentFirst;
  }
}

enum UserMessageBubbleStyle {
  followCardOpacity,
  fixedGray,
}

extension UserMessageBubbleStyleX on UserMessageBubbleStyle {
  String get storageKey => switch (this) {
        UserMessageBubbleStyle.followCardOpacity => 'follow_card_opacity',
        UserMessageBubbleStyle.fixedGray => 'fixed_gray',
      };

  String get label => switch (this) {
        UserMessageBubbleStyle.followCardOpacity => '跟随卡片透明度',
        UserMessageBubbleStyle.fixedGray => '固定灰底',
      };

  static UserMessageBubbleStyle fromStorage(String? value) {
    for (final style in UserMessageBubbleStyle.values) {
      if (style.storageKey == value) return style;
    }
    return UserMessageBubbleStyle.followCardOpacity;
  }
}

class TransactionCardText {
  final String title;
  final String secondary;

  const TransactionCardText({required this.title, this.secondary = ''});
}

TransactionCardText resolveTransactionCardText({
  required TransactionCardDisplayMode mode,
  required TransactionKind kind,
  required String note,
  required String categoryName,
  String accountName = '',
  String toAccountName = '',
}) {
  final cleanNote = note.trim();
  final cleanCategory = categoryName.trim().isNotEmpty
      ? categoryName.trim()
      : kind == TransactionKind.income
          ? '其他收入'
          : '未分类';

  if (kind == TransactionKind.transfer) {
    final from = accountName.trim();
    final to = toAccountName.trim();
    final route = from.isNotEmpty && to.isNotEmpty
        ? '$from → $to'
        : (from.isNotEmpty ? from : (to.isNotEmpty ? to : '转账'));
    return TransactionCardText(title: route, secondary: cleanNote);
  }

  if (mode == TransactionCardDisplayMode.contentFirst && cleanNote.isNotEmpty) {
    return TransactionCardText(title: cleanNote, secondary: cleanCategory);
  }
  return TransactionCardText(
    title: cleanCategory,
    secondary: cleanNote,
  );
}

String transactionCardTimeLabel(
  DateTime date, {
  required bool dateGrouped,
  required TransactionTimePrecision precision,
}) {
  String two(int value) => value.toString().padLeft(2, '0');
  final showClock = shouldShowTransactionClock(date, precision);
  final time = '${two(date.hour)}:${two(date.minute)}';
  if (dateGrouped) return showClock ? time : '';
  final day = '${date.year}/${two(date.month)}/${two(date.day)}';
  return showClock ? '$day $time' : day;
}

String joinTransactionCardDetails(Iterable<String?> parts) => parts
    .map((part) => part?.trim() ?? '')
    .where((part) => part.isNotEmpty)
    .join(' · ');

File: ios-app/QingJiCore/Sources/QingJiCore/Models/TransactionCardDisplay.swift
import Foundation

/// 账单卡片的标题优先级。与 Android 的 transaction_card_display.dart
/// 使用相同的持久化值，iOS 只替换呈现控件。
public enum TransactionCardDisplayMode: String, Codable, CaseIterable, Hashable, Identifiable, Sendable {
    case contentFirst = "content_first"
    case categoryFirst = "category_first"

    public var id: String { rawValue }

    public var label: String {
        switch self {
        case .contentFirst: return "内容优先"
        case .categoryFirst: return "分类优先"
        }
    }
}

/// 喵助手用户消息气泡的背景策略。
public enum UserMessageBubbleStyle: String, Codable, CaseIterable, Hashable, Identifiable, Sendable {
    case followCardOpacity = "follow_card_opacity"
    case fixedGray = "fixed_gray"

    public var id: String { rawValue }

    public var label: String {
        switch self {
        case .followCardOpacity: return "跟随卡片透明度"
        case .fixedGray: return "固定灰底"
        }
    }
}

public struct TransactionCardText: Equatable, Sendable {
    public let title: String
    public let secondary: String

    public init(title: String, secondary: String = "") {
        self.title = title
        self.secondary = secondary
    }
}

/// 同一条流水在 Android/iOS 上采用相同的信息层级，避免只因换平台就
/// 让用户先看到不同的字段。
public func resolveTransactionCardText(
    mode: TransactionCardDisplayMode,
    kind: TransactionKind,
    note: String,
    categoryName: String,
    accountName: String = "",
    toAccountName: String = ""
) -> TransactionCardText {
    let cleanNote = note.trimmingCharacters(in: .whitespacesAndNewlines)
    let cleanCategory: String
    if !categoryName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        cleanCategory = categoryName.trimmingCharacters(in: .whitespacesAndNewlines)
    } else {
        cleanCategory = kind == .income ? "其他收入" : "未分类"
    }

    if kind == .transfer {
        let from = accountName.trimmingCharacters(in: .whitespacesAndNewlines)
        let to = toAccountName.trimmingCharacters(in: .whitespacesAndNewlines)
        let route: String
        if !from.isEmpty && !to.isEmpty {
            route = "\(from) → \(to)"
        } else if !from.isEmpty {
            route = from
        } else if !to.isEmpty {
            route = to
        } else {
            route = "转账"
        }
        return TransactionCardText(title: route, secondary: cleanNote)
    }

    if mode == .contentFirst && !cleanNote.isEmpty {
        return TransactionCardText(title: cleanNote, secondary: cleanCategory)
    }
    return TransactionCardText(title: cleanCategory, secondary: cleanNote)
}

/// 日期分组账单行的附加时间文本。日期精度不足时不伪造时分。
public func transactionCardTimeLabel(
    _ date: Date,
    dateGrouped: Bool,
    precision: TransactionTimePrecision
) -> String {
    let showClock = precision == .exact || precision == .entryClock
    let formatter = DateFormatter()
    formatter.locale = Locale(identifier: "zh_CN")
    formatter.dateFormat = dateGrouped ? "HH:mm" : (showClock ? "yyyy/MM/dd HH:mm" : "yyyy/MM/dd")
    return dateGrouped && !showClock ? "" : formatter.string(from: date)
}

public func joinTransactionCardDetails(_ parts: [String]) -> String {
    parts.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        .filter { !$0.isEmpty }
        .joined(separator: " · ")
}
